nimbus initial commit

pedromiguelmiranda opened the pull request:
project structure with:
- chronicles support
- config files 
- layers handlers.

arnetheduck commented on Makefile:
```diff
@@ -104,6 +104,7 @@ VERIF_PROXY_OUT_PATH ?= build/libverifproxy/
 	update \
 	nimbus \
 	nimbus_execution_client \
+	nimbus_unified \
```
why not name it nimbus?

pedromiguelmiranda replied:
the target `nimbus` is on rampdown:
https://github.com/status-im/nimbus-eth1/blob/dev/pedro/kikof/Makefile#L217
![image](https://github.com/user-attachments/assets/7025d183-ca08-415c-9097-d730518e8ae3)

and we need to solve issue #2738 and update some README `make` commands [refs](https://github.com/status-im/nimbus-eth1/blob/master/README.md?plain=1#L59)

Given that this is a project in development yet to be released (https://nimbus.guide/execution-client.html), how much time should we wait for the rampdown period?

Agree that it would be nice to get the renaming before starting landing this.

pedromiguelmiranda replied:
Fixed in 411a3cadfa3ac0be5eaf4b7d69b863c169d75957

arnetheduck commented on nimbus_unified/tests/test_nimbus_unified.nim:
```diff
@@ -0,0 +1,127 @@
+    check isShutDownRequired.load() == true
+
+    #services running should be 0
+    for i in 0 .. cNimbusMaxServices - 1:
```
in all of these cases, you can `for s in nimbus.serviceList: check s.isNone`  - no need for a constant

arnetheduck commented on nimbus_unified/nimbus_unified.nim:
```diff
@@ -0,0 +1,155 @@
+  notice "Exited all services"
+
+## Service monitoring
+proc monitor*(nimbus: Nimbus) =
```
what's this for, ie what will it be doing?

pedromiguelmiranda replied:
Monitor if one of the services ended (gracefully) and an to ask remaining services to end gracefully.
(check one the service prototype wrappers for an example).

This is a small naive mechanism using polling, to be replaced by some wake up mechanism (maybe a condition?)

arnetheduck commented on nimbus_unified/version.nim:
```diff
@@ -0,0 +1,62 @@
+  # Short debugging identifier to be placed in the ENR
+  enrClientInfoShort* = toBytes("f")
+
+func getNimGitHash*(): string =
```
we have a "nimbus_binary_common" file for hosting little helpers like this - the idea with that file is no matter what the binary contains, certain things work the same (logging options, version numbers etc) - it would be good to refactor the existing binaries to make use of such utilities, just like the logging.

pedromiguelmiranda replied:
Great. all projects are implementing similar code for the same purpous.
Let me try to land that before this.

arnetheduck commented on nimbus_unified/tests/test_nimbus_unified.nim:
```diff
@@ -0,0 +1,127 @@
+# Helper Functions
+# ----------------------------------------------------------------------------
+
+template fileExists(filename: string): bool =
```
what's wrong with https://nim-lang.org/docs/files.html#fileExists%2CPath?

arnetheduck replied:
as an alternative, if this function does not exist in `io2`, it's better to add a correctly implemented function there (ie an implementation that uses the right os utilities, similar to the above linked documentation but with different error handling

pedromiguelmiranda replied:
absolutely nothing wrong with those components. This was my half way implementing unit tests for the `createPidFile` which is not being used.
Which will be moved to binary common as well

arnetheduck commented on nimbus_unified/tests/test_nimbus_unified.nim:
```diff
@@ -0,0 +1,127 @@
+template removeFile(filename: string) =
+  try:
+    discard io2.removeFile(filename)
+  except IOError:
```
`io2` does not raise exceptions, or at least it shouldn't - we use either `Result` or exceptions but not both in a single function (with a few exceptions that don't apply here).

pedromiguelmiranda replied:
same as preivous comment

pedromiguelmiranda commented on nimbus_unified/nimbus_unified.nim:
```diff
@@ -0,0 +1,155 @@
+  setControlCHook(controlCHandler)
+
+  var
+    execution = LayerConfig(kind: Execution, executionConfig: NimbusConf())
```
This is a nice hint. thank you
Wasn't aware of the configs..

pedromiguelmiranda commented:
latest updates:
- removed dead code.
- removed some structures and functionalitires not needed yet
- thread communication with shared data via channels: instead of creating each client configuration and pass itto each service thread, an array of configs will be passed instead. Each service thread well haev the responsability of creating its configuration and manage it.

tersec commented on nimbus/tests/nim.cfg:
```diff
@@ -0,0 +1,14 @@
+
+--styleCheck:usages
+--styleCheck:hint
+--hint[Processing]:off
```
Why the apparently random mix of files ending without newlines and ones which do?

pedromiguelmiranda replied:
great find. 
the project is using the nph formatter which formats all the nim files, and I forgot the non nim ones.
